java/ql/lib/semmle/code/java/frameworks/apache/Collections.qll

```diff
@@ -182,6 +204,695 @@ private class ApacheKeyValueModel extends SummaryModelCsv {
+          ".iterators;ZippingIterator;true;ZippingIterator;(Iterator[]);;Element of ArrayElement of Argument[0];Element of Argument[-1];value",
+          ".iterators;ZippingIterator;true;ZippingIterator;(Iterator,Iterator);;Element of Argument[0];Element of Argument[-1];value",
+          ".iterators;ZippingIterator;true;ZippingIterator;(Iterator,Iterator);;Element of Argument[1];Element of Argument[-1];value",
+          ".iterators;ZippingIterator;true;ZippingIterator;(Iterator,Iterator,Iterator);;Element of Argument[0];Element of Argument[-1];value",
+          ".iterators;ZippingIterator;true;ZippingIterator;(Iterator,Iterator,Iterator);;Element of Argument[1];Element of Argument[-1];value",
+          ".iterators;ZippingIterator;true;ZippingIterator;(Iterator,Iterator,Iterator);;Element of Argument[2];Element of Argument[-1];value"
+        ]
+  }
+}
+
+/**
+ * Value-propagating models for the package `org.apache.commons.collections4.list`.
+ */
+private class ApacheListModel extends SummaryModelCsv {
+  override predicate row(string row) {
+    row =
+      ["org.apache.commons.collections4", "org.apache.commons.collections"] +
+        [
+          // Note that when lambdas are supported we should have more models for TransformedList
+          ".list;AbstractLinkedList;true;AbstractLinkedList;;;Element of Argument[0];Element of Argument[-1];value",
+          ".list;AbstractLinkedList;true;getFirst;;;Element of Argument[-1];ReturnValue;value",
+          ".list;AbstractLinkedList;true;getLast;;;Element of Argument[-1];ReturnValue;value",
+          ".list;AbstractLinkedList;true;addFirst;;;Argument[0];Element of Argument[-1];value",
+          ".list;AbstractLinkedList;true;addLast;;;Argument[0];Element of Argument[-1];value",
+          ".list;AbstractLinkedList;true;removeFirst;;;Element of Argument[-1];ReturnValue;value",
+          ".list;AbstractLinkedList;true;removeLast;;;Element of Argument[-1];ReturnValue;value",
+          ".list;AbstractLinkedList;true;updateNode;;;Argument[1];Element of Argument[-1];value",
+          ".list;AbstractLinkedList;true;updateNode;;;Argument[1];Element of Argument[0];value",
+          ".list;AbstractLinkedList;true;createNode;;;Argument[0];Element of ReturnValue;value",
+          ".list;AbstractLinkedList;true;addNodeBefore;;;Argument[1];Element of Argument[-1];value",
+          ".list;AbstractLinkedList;true;addNodeBefore;;;Argument[1];Element of Argument[0];value",
+          ".list;AbstractLinkedList;true;addNodeAfter;;;Argument[1];Element of Argument[-1];value",
+          ".list;AbstractLinkedList;true;addNodeAfter;;;Argument[1];Element of Argument[0];value",
+          ".list;AbstractLinkedList;true;addNode;;;Element of Argument[0];Element of Argument[-1];value",
+          ".list;AbstractLinkedList;true;addNode;;;Element of Argument[0];Element of Argument[1];value",
```

Comment:
Not sure how this affects `Argument[1]`. From the code, it does touch the linked nodes next/previous fields but that's no value flow

Reply:
See :fish: